base/abstractarray.jl

```diff
@@ -2506,6 +2506,237 @@ end
+_stack(dims::Union{Integer, Colon}, iter) = _stack(dims, IteratorSize(iter), iter)
+
+_stack(dims, ::IteratorSize, iter) = _stack(dims, collect(iter))
+
+function _stack(dims, ::Union{HasShape, HasLength}, iter)
+    S = @default_eltype iter
+    T = S != Union{} ? eltype(S) : Any  # Union{} occurs for e.g. stack(1,2), postpone the error
+    if isconcretetype(T)
+        _typed_stack(dims, T, S, iter)
+    else  # Need to look inside, but shouldn't run an expensive iterator twice:
+        array = iter isa Union{Tuple, AbstractArray} ? iter : collect(iter)
+        isempty(array) && return _empty_stack(dims, T, S, iter)
+        T2 = mapreduce(eltype, promote_type, array)
+        _typed_stack(dims, T2, eltype(array), array)
+    end
+end
+
+function _typed_stack(::Colon, ::Type{T}, ::Type{S}, A, Aax=_axes(A)) where {T, S}
+    xit = iterate(A)
+    nothing === xit && return _empty_stack(:, T, S, A)
+    x1, _ = xit
+    ax1 = _axes(x1)
+    B = similar(_first_array(x1, A), T, ax1..., Aax...)
+    off = firstindex(B)
+    len = length(x1)
+    while xit !== nothing
+        x, state = xit
+        _stack_size_check(x, ax1)
+        copyto!(B, off, x)
+        off += len
+        xit = iterate(A, state)
+    end
+    B
+end
+
+_axes(x) = _axes(x, IteratorSize(x))
+_axes(x, ::HasLength) = (OneTo(length(x)),)
+_axes(x, ::IteratorSize) = axes(x)
+
+# For some dims values, stack(A; dims) == stack(vec(A)), and the : path will be faster
+_typed_stack(dims::Integer, ::Type{T}, ::Type{S}, A) where {T,S} =
+    _typed_stack(dims, T, S, IteratorSize(S), A)
+_typed_stack(dims::Integer, ::Type{T}, ::Type{S}, ::HasLength, A) where {T,S} =
+    _typed_stack(dims, T, S, HasShape{1}(), A)
+function _typed_stack(dims::Integer, ::Type{T}, ::Type{S}, ::HasShape{N}, A) where {T,S,N}
+    if dims == N+1
+        _typed_stack(:, T, S, A, (_vec_axis(A),))
+    else
+        _dim_stack(dims, T, S, A)
+    end
+end
+_typed_stack(dims::Integer, ::Type{T}, ::Type{S}, ::IteratorSize, A) where {T,S} =
+    _dim_stack(dims, T, S, A)
+
+_vec_axis(A, ax=_axes(A)) = length(ax) == 1 ? only(ax) : OneTo(prod(length, ax; init=1))
+
+function _dim_stack(dims::Integer, ::Type{T}, ::Type{S}, A) where {T,S}
+    xit = iterate(A)
+    nothing === xit && return _empty_stack(dims, T, S, A)
+    x1, _ = xit
+    ax1 = _axes(x1)
+    N1 = length(ax1)+1
+    dims in 1:N1 || throw(ArgumentError("cannot stack slices ndims(x) = $(N1-1) along dims = $dims"))
+
+    newaxis = _vec_axis(A)
+    outax = ntuple(d -> d==dims ? newaxis : _axes(x1)[d - (d>dims)], N1)
+    B = similar(_first_array(x1, A), T, outax...)
+
+    iit = iterate(newaxis)
+    while xit !== nothing
+        x, state = xit
+        i, istate = iit
+        _stack_size_check(x, ax1)
+        @inbounds if dims==1
+            inds1 = ntuple(d -> d==1 ? i : Colon(), N1)
+            if x isa AbstractArray
+                B[inds1...] = x
+            else
+                copyto!(view(B, inds1...), x)
+            end
+        else
```

Comment:
I guess L2677~L2681 is for performance reason?
If so, it might be good to leave a `TODO` here for future code clean after PR like #38014?
(It so annoy that they have performance difference)

Reply:
How about split the above `while` loop to `_dim_stack!` and use `Val` to make it stable?
```julia
    dims == 1 && return _dim_stack!(Val(1), B, ax1, A, xit, newaxis)
    dims == 2 && return _dim_stack!(Val(2), B, ax1, A, xit, newaxis)
    return _dim_stack!(Val(dims), B, ax1, A, xit, newaxis)
end

function _dim_stack!(::Val{dims}, B, ax1, A, xit, newaxis) where {dims}
    iit = iterate(newaxis)
    @inbounds while xit !== nothing
        x, state = xit
        i, istate = iit
        _stack_size_check(x, ax1)
        inds = ntuple(d -> d==dims ? i : Colon(), ndims(B))
        copyto!(view(B, inds...), x)
        xit = iterate(A, state)
        iit = iterate(newaxis, istate)
    end
    B
end
```

Reply:
Yes to performance; I am not sure it's a SIMD thing like #38014, I thought it was more that the type of the `view` depends on `dims` and this throws something off. But that might be my superstitions.

I tried `_dim_stack!(::Val{dims}` and never got it to play nicely with non-Val `stack(...; dims)`. Maybe calling it from manual dispatch would help. Notice that in the present code, any `dims` is fine for an iterator of arrays, it is not limited to 2. The special-casing is allows for an iterator of tuples with `dims=1` (this path) or `dims>1` (the colon path). To hit the unstable view path, you need an iterator of non-arrays which themselves have `ndims>1`, and `1<dims<ndims(out)`.

Reply:
I tried 095050a53aa16bb6df46e1165b112db67c143987. And the following seems work well:
```julia
@btime stack(reshape(1:1000,10,10,10) for _ = 1:1000; dims = 2)
  2.030 ms (2 allocations: 7.63 MiB)
@btime stack(reshape(1:1000,10,10,10) for _ = 1:1000; dims = 3)
  1.371 ms (2 allocations: 7.63 MiB)
@btime stack(reshape(1:1000,10,10,10) for _ = 1:1000; dims = 4) # for compare
  1.132 ms (3 allocations: 7.63 MiB)
```

My `copyto_unaliased` has some local tuning, and it's faster than `setindex!` for most cases.

Reply:
Thanks, will have a look in a bit. My tests were https://github.com/JuliaLang/julia/pull/43334#discussion_r771980548, I guess you are using the path used there for tuples for arrays too.

Reply:
With my computer, today's master, this idea speeds up your test nicely. But it causes one of mine to allocate:
```julia
julia> @btime stack(reshape(1:1000,10,10,10) for _ = 1:1000; dims = 2);
  min 4.694 ms, mean 5.482 ms (36426 allocations, 8.63 MiB)  # PR before
  min 2.079 ms, mean 2.283 ms (2 allocations, 7.63 MiB)  # with this Val(dims)

julia> xv = [rand(10) for _ in 1:100];

julia> @btime stack($xv; dims=1);
  min 1.458 μs, mean 2.601 μs (1 allocation, 7.94 KiB)  # PR before
  min 1.983 μs, mean 3.256 μs (100 allocations, 11.03 KiB) # with this Val(dims)
```

Reply:
force inline `dims = 1` fix the allocation on my PC.
`@inline dims == 1 && return _dim_stack!(Val(1), B, ax1, A, xit, newaxis)`
Hope this also works for you.